### DISCUSSION TITLE
prevent non-latin chars in slug for existing entries (limitAutoSlugsToAscii)

### DISCUSSION THREAD

**Discussion by mfell** [craftcms/cms#13860] [Ideas] [answered]
My suggestion is with reference to [Docs: limitAutoSlugsToAscii](https://craftcms.com/docs/4.x/config/general.html#limitautoslugstoascii)  
That works fine ... no bug or issue here.

My "problem" is, that I'm German and German editors really like **German Umlauts** (**äöüß**).  
_(Same topic in other languages, of course!)_

### Steps to reproduce

1. Create a new entry with a title which contains Umlauts, f.e. `häßliches Entlein`.
2. Save entry -> Craft limitAutoSlugsToAscii -> the slug is now `haessliches-entlein`. perfect!
3. A German editor thinks _"That looks wrong."_ and changes it back to `häßliches-Entlein`.
4. Save entry -> slug (URL) contains Umlauts `häßliches-entlein` ... I don't want this in slug.

![Bildschirmfoto 2023-10-20 um 15 00 41](https://github.com/craftcms/cms/assets/6691773/eafa6323-7135-403b-80f9-c359c6a33d92)

### Suggestion

Add a config option to Craft like `limitAutoSlugsToAsciiOnSave`?  
Just the same functionality and configuration like limitAutoSlugsToAscii, but triggering after saving.

The possibility for editors to modify the slug is fine ... I only want to get rid of German Umlauts.

**Comment by brandonkelly** [comment, marked as answer]
The setting includes the word `Auto` because it’s intentionally only concerned with _automatically_-generated slugs. 

If you want to completely prevent non-ASCII characters in slugs, you could write a module that hooks into element saving, and ensures non-ASCII characters are replaced before validation.

```php
use craft\base\Element;
use craft\base\ElementInterface;
use craft\base\Event;
use craft\events\ModelEvent;
use craft\helpers\StringHelper;

Event::on(Element::class, Element::EVENT_BEFORE_SAVE, function(ModelEvent $event) {
    /** @var ElementInterface $element */
    $element = $event->sender;

    if ($element->slug) {
        $element->slug = StringHelper::toAscii($element->slug, $element->getSite()->language);
    }
});
```

**Reply by mfell** [reply]
I see ... thanks a lot, Brandon!
